pkg/mesh/td.go

```diff
@@ -0,0 +1,227 @@
+	hostName, hostErr := os.Hostname()
+	if hostErr != nil {
+		return "", hostErr
+	}
+	ips, ipErr := net.LookupIP(hostName)
+	if ipErr != nil {
+		return "", ipErr
+	}
+	return fmt.Sprintf("%s~%s", uuid.New().String(), ips[0].String()), nil
+}
+
+// Prepare TdSidecarEnv by using Metadata server and environment variables.
+func (td *TdSidecarEnv) prepare() {
+	// We use the project number that this cloud run service is running under as a default.
+	if projectNumber, err := td.fetchProjectNumber(); err != nil {
+		log.Println("Unable to auto-generate project_number.")
+	} else {
+		td.ProjectNumber = projectNumber
+	}
+
+	td.loadFromEnv()
+
+	if nodeID, err := td.fetchNodeID(); err != nil {
+		td.NodeID = fmt.Sprintf("%s~%s", uuid.New().String(), "127.0.0.1")
+		log.Println("Unable to generate proper nodeID, using: ", td.NodeID)
+	} else {
+		td.NodeID = nodeID
+	}
+
+	if zone, err := td.fetchZone(); err != nil {
+		td.EnvoyZone = "cloud-run-cluster"
+		log.Println("Unable to generate proper zone info, using: ", td.EnvoyZone)
+	} else {
+		td.EnvoyZone = zone
+	}
+}
+
+func (td *TdSidecarEnv) getIPTablesInterceptionEnvVars() []string {
+	envs := []string{
+		"TRAFFIC_DIRECTOR_GCE_VM_DEPLOYMENT_OVERRIDE=true",
+		"DISABLE_REDIRECTION_ON_LOCAL_LOOPBACK=true",
+		fmt.Sprintf("%s=%s", "ENVOY_DNS_PORT", td.EnvoyDnsPort),
+	}
+	return envs
+}
+
+func (td *TdSidecarEnv) logLevel() string {
+	return td.LogLevel
+}
+
+func (td *TdSidecarEnv) envoyPort() string {
+	return td.EnvoyPort
+}
+
+func (td *TdSidecarEnv) envoyAdminPort() string {
+	return td.EnvoyAdminPort
+}
+
+func (td *TdSidecarEnv) projectNumber() string {
+	return td.ProjectNumber
+}
+
+func (td *TdSidecarEnv) networkName() string {
+	return td.NetworkName
+}
+
+func (td *TdSidecarEnv) serviceCIDR() string {
+	return td.ServiceCidr
+}
+
+func (td *TdSidecarEnv) envoyLogDirectory() string {
+	return td.LogDirectory
+}
+
+// Reads bootstrap template found in absolute path templatePath and replaces strings.
+func (td *TdSidecarEnv) prepareTrafficDirectorBootstrap(templatePath string, outputPath string) error {
+	data, err := os.ReadFile(templatePath)
+	if err != nil {
+		return err
+	}
+
+	template := string(data)
+	template = strings.ReplaceAll(template, "ENVOY_NODE_ID", td.NodeID)
+	template = strings.ReplaceAll(template, "ENVOY_ZONE", td.EnvoyZone)
+	template = strings.ReplaceAll(template, "VPC_NETWORK_NAME", td.NetworkName)
+	template = strings.ReplaceAll(template, "CONFIG_PROJECT_NUMBER", td.ProjectNumber)
+	template = strings.ReplaceAll(template, "ENVOY_PORT", td.EnvoyPort)
+	template = strings.ReplaceAll(template, "ENVOY_ADMIN_PORT", td.EnvoyAdminPort)
+	template = strings.ReplaceAll(template, "XDS_SERVER_CERT", td.XdsServerCert)
+	template = strings.ReplaceAll(template, "TRACING_ENABLED", strconv.FormatBool(td.TracingEnabled))
+	template = strings.ReplaceAll(template, "ACCESSLOG_PATH", "")
+	template = strings.ReplaceAll(template, "BACKEND_INBOUND_PORTS", "")
```

Comment:
Right now CloudRun is only using 8080 - we may support multiple ports at some point, using the tunnel - but we'll need to figure out what API to use, since port type is also neeed.